src/main/java/org/truffleruby/interop/ForeignReadStringCachedHelperNode.java

```diff
@@ -40,15 +39,28 @@
     @Child private CallDispatchHeadNode callNode;
 
     protected final static String INDEX_METHOD_NAME = "[]";
+    protected final static String METHOD_NAME = "method";
 
     public abstract Object executeStringCachedHelper(VirtualFrame frame, DynamicObject receiver, Object name, Object stringName, boolean isIVar);
 
-    @Specialization(guards = "isIVar")
+    @Specialization(guards = "isRubyArray(receiver) || isRubyHash(receiver)")
+    public Object readArrayHash(
+            VirtualFrame frame,
+            DynamicObject receiver,
+            Object name,
+            Object stringName,
+            boolean isIVar,
+            @Cached("create()") ForeignToRubyNode nameToRubyNode) {
+        return getCallNode().call(frame, receiver, INDEX_METHOD_NAME, nameToRubyNode.executeConvert(name));
+    }
+
+    @Specialization(guards = {"!isRubyArray(receiver)", "!isRubyHash(receiver)", "isIVar", "stringName.equals(cachedStringName)"})
```

Comment:
I think we should have a node reading ivars with the name as an argument, so the caching is done there. There might already be a node doing that. The advantage is the caching can be done better, e.g. trying `==` first.

Reply:
It's `ObjectIVarGetNode` - but actually this caches pretty badly. We should fix this...